internal/metadataproviders/aws/eks/metadata.go

```diff
@@ -0,0 +1,171 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package eks // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/metadataproviders/aws/eks"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	clusterNameAwsEksTag     = "aws:eks:cluster-name"
+	clusterNameEksTag        = "eks:cluster-name"
+	kubernetesClusterNameTag = "kubernetes.io/cluster/"
+)
+
+type Provider interface {
+	AccountID(ctx context.Context) (string, error)
+	ClusterName(ctx context.Context, instanceID string) (string, error)
+	ClusterVersion() (string, error)
+	InstanceID() string
+	AvailabilityZone() string
+	Region() string
+	InitializeInstanceMetadata(ctx context.Context, nodeName string) error
+}
+
+type metadataClient struct {
+	instanceMetadata instanceMetadata
+	clientset        *kubernetes.Clientset
+	ec2Client        *ec2.Client
+	stsClient        *sts.Client
+}
+
+type instanceMetadata struct {
+	availabilityZone string
+	instanceID       string
+	region           string
+}
+
+var _ Provider = (*metadataClient)(nil)
+
+func NewProvider(cfg aws.Config) (Provider, error) {
+	conf, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize provider. Could not load cluster config: %w", err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(conf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize provider. Could not create clientset for Kubernetes client: %w", err)
+	}
+
+	return &metadataClient{
+		clientset: clientset,
+		ec2Client: ec2.NewFromConfig(cfg),
+		stsClient: sts.NewFromConfig(cfg),
+	}, nil
+}
+
+// InitializeInstanceMetadata should be called right after NewProvider to:
+// 1. Get the instanceID which is needed for the ec2 describe call
+// 2. Get the realm which is needed to pin the ec2 and sts call to a specific region.
+func (c *metadataClient) InitializeInstanceMetadata(ctx context.Context, nodeName string) error {
+	if nodeName == "" {
+		return errors.New("node name is not set")
+	}
+
+	node, err := c.clientset.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to retrieve node %w", err)
+	}
+
+	region, availabilityZone, instanceID := parseRegionAndInstanceID(node.Spec.ProviderID)
+	// attributes needed for the remaining calls to work
+	if region == "" || instanceID == "" {
+		return fmt.Errorf("failed to retrieve region or instanceId from providerID: %s", node.Spec.ProviderID)
+	}
+
+	c.instanceMetadata = instanceMetadata{availabilityZone: availabilityZone, instanceID: instanceID, region: region}
+	return nil
+}
+
+func parseRegionAndInstanceID(input string) (string, string, string) {
```

Comment:
I think it would be good to add some test for this to ensure we don't break parsing logic

Reply:
Thanks! for sure, tests will be added